README.md
# pe_support_script

#### Table of Contents

1. [Module Description - What the module does and why it is useful](#module-description)
2. [Setup - The basics of getting started with pe_support_script](#setup)
3. [Usage - Configuration options and additional functionality](#usage)
4. [Reference - An under-the-hood peek at what the module is doing and how](#reference)
5. [Limitations - OS compatibility, etc.](#limitations)
6. [Development - Guide for contributing to the module](#development)


## Module Description

This module provides the `puppet enterprise support` face which is used
to gather a snapshot of PE configuration, log messages and other diagnostic
data that is essential to troubleshooting PE infrastructure.


## Setup

Place the pe_support_script module on the `$basemodulepath` of the
Master of Masters (MoM) in a PE installation. The module will then
be distributed to all PE nodes via pluginsync.

The configuration described above is produced by the PE Installer and
should only need to be done manually in non-production circumstances
such as acceptance tests.


## Usage

The support script is used by invoking the `puppet enterprise support`
subcommand as the `root` user. The script will print out a list of
data gathered and finishes by printing the location of a tarball
containing the output:

```
# /opt/puppetlabs/bin/puppet enterprise support
Puppet Enterprise Support Script v1.3.0
Creating drop directory at /var/tmp/puppet_enterprise_support_pe-201643-master_20170206144746
Collecting information

 ** Collecting output of: netstat -anptu

 ** Collecting output of: /usr/sbin/sestatus

...

Please submit /var/tmp/puppet_enterprise_support_pe-201643-master_20170206144746.tar.gz to Puppet Support using the upload site you've been invited to.
```


## Reference

The `puppet enterprise support` face operates by execcing a bash script
located at:

 - [lib/puppet_x/puppetlabs/support_script/v1/puppet-enterprise-support.sh](lib/puppet_x/puppetlabs/support_script/v1/puppet-enterprise-support.sh)

This script is laid out in three general sections:

  - Constants and utility functions.
  - Diagnostic functions that run diagnostics and collect data.
  - A main section which sets up the script output and runs a
    list of diagnostic functions.

The recommended approach for familiarizing yourself with the support
script code is to start at the end with the main section and then
search up for the implementation of diagnostic functions of interest.

The [docs/archive_layout.md](docs/archive_layout.md) file contains
an overview of the files included in support script output.


## Limitations

The `puppet enterprise support` command was developed to aid the
troubleshooting of PE infrastructure nodes --- MoMs, databases, consoles,
compile masters, MCollective hubs and spokes, etc. Therefore, this module
only supports Linux OS versions listed as "Puppet master platforms" for a
given PE release.

The list for the 2018.1 LTS series is at:

https://puppet.com/docs/pe/2018.1/supported_operating_systems.html#puppet-master-platforms

Support for troubleshooting Windows agents was added in v2.9.0 of the support
script which first shipped with PE 2018.1.5 and 2019.0.0.

## Development

The easiest way to contribute to the development of the PE Support Script
is to file tickets for bugs or improvements. Tickets should be filed
in the https://tickets.puppet.com/browse/PE tracker with the component
field set to "Support Script". The component bit is **very important** as
a Support Script ticket will just disappear into the JIRA backlog without it.

### Adding to the Script

The business end of the support script is currently implemented as a single
bash script as described in the [reference section](#reference). Diagnostics
can be extended by adding to the relevant bash functions and new diagnostics
can be added by creating new functions in the middle of the script and
adding a call to them into the main section at the end of the script.

Keep in mind the following guidelines when developing the support script:

  - Keep it safe. The script should default to read-only operations and
    should avoid collecting data that may contain sensitive information
    such as passwords or encryption keys. Running the support script
    should not modify a PE installation in any way beyond copying data
    to the output directory under `/var/tmp`.

  - New additions to the bash support script should follow the
    [Shell Style Guide][shell-guide] and pass a lint check administered by
    [Shellcheck][shellcheck].

  - Diagnostic functions should include guard statements that turn them
    into no-ops if required commands or data are missing. For example, the
    script shouldn't attempt diagnostics related to systemd if the `systemctl`
    command is not present. This keeps the script performant, and maximizes
    compatiblity accross OS versions and PE versions.

  - Structure your development work to produce a single commit per logical
    change. For example: if you extend an existing diagnostic, add a new
    diagnostic, and perform some stylistic cleanup, then each of those changes
    should be a separate commit.

  - Development should be done against the LTS branch of the support script
    unless functionality is being added that is only useful for newer PE
    versions.

[shell-guide]: https://google.github.io/styleguide/shell.xml
[shellcheck]: https://www.shellcheck.net/

### Testing

The support script is currently tested using a Beaker acceptance suite
which runs the script against a PE installation on each supported master
platform. ["Bash Strict Mode"][strict-mode] is enabled when the script is
run as part of an acceptance test. This extra layer of checking catches
the use of undeclared variables or diagnostic functions that are failing
due to incorrect input or missing guard statements.

To run the acceptance suite, first connect to the Puppet VPN as the test
suite makes use of several resources that are only available on the internal
network.

Next, use Bundler to install dependencies such as Beaker:

    bundle install --path=.bundle/lib

And ensure that a copy of the [acceptance testing key][acceptance-key]
is available at:

    ~/.ssh/id_rsa-acceptance

Then, execute the test helper under the `ext/` directory:

    ./ext/run_acceptance_tests.sh

The wrapper defaults to testing against the latest 2018.1 LTS nightly build.
Other PE builds can be tested by passing an X.Y version number, or the string
"master" to the helper:

    ./ext/run_acceptance_tests.sh master

The helper will launch parallel tests for every PE configuration described
by files in the tests/beaker/configs/ directory. Tests can take over ten
minutes to run, but when they've finished a Ruby webserver will be launched
to display the results:

    http://localhost:8000

[strict-mode]: http://redsymbol.net/articles/unofficial-bash-strict-mode/
[acceptance-key]: https://confluence.puppetlabs.com/display/QE/SSH+access+to+vmpooler+VMs


CHANGELOG.md
## [Release 2.13.2] - 2020-04-03

Bugfix release of the Support Script.

### Bug Fixes

  - Ruby support script now collects service configuration from `/etc/default`
    and `/etc/sysconfig`.

  - Various and sundry fixes and improvements to documentation and test suites.


## [Release 2.13.1] - 2020-01-13

Bugfix release of the Support Script.

### Bug Fixes

  - Ruby support script now applies `gzip` compression to large output files.

  - Ruby support script no longer prints exit code to `stdout`.

  - Ruby support script now collects the size of Postgres TOAST tables.

  - Ruby support script now runs on RedHat variants such as Oracle Enterprise Linux
    and Scientific Linux.

  - Ruby support script responds to Ctrl-C, Ctrl-\, and SIGTERM.


## [Release 2.13.0] - 2019-10-16

### Summary

Feature and bugfix release of the Support Script.

### Features

  - Support for PE 2019.2

  - Support for RedHat 7 FIPS masters

  - Large re-write of the `--v3` Ruby version of the support script.
    This updated script groups diagnostics into scopes centered around
    PE Services that can be toggled on and off with the new `--enable`,
    `--disable`, and `--only` flags.


### Bug Fixes

  - The bash version of the support script properly collects disk space
    data from Postgres 11


## [Release 2.12.0] - 2019-07-17

### Summary

Feature and bugfix release of the Support Script.

### Features

  - Configuration and log files for the PE 2019.1 `pe-ace-server` service
    are collected from `/etc/puppetlabs/ace-server/conf.d` and from
    `/var/log/puppetlabs/ace-server`.

  - The size of the filebucket upload directory, `/opt/puppetlabs/server/data/puppetserver/bucket`
    is collected.

  - The `dnf.conf` file is collected on RedHat 8 systems.

  - Human-readable date stamps are added to `dmesg` output when available.

### Bug Fixes

  - The Ruby version of the support script now properly collects `gem list`
    output.


## [Release 2.11.0] - 2019-04-17

### Summary

Feature and bugfix release of the Support Script.

### Features

  - The Ruby support script now has an `--upload` option that uploads results
    to SFTP servers hosted by Puppet Inc. Requires the `--v3` flag to be
    passed and a value to be specified for the `--ticket` flag.

  - The support script now collects the output of `puppet infrastructure tune`.

  - The support script now collects process information from `/proc` for each
    PE service.

  - The support script now collects information from `/sys/fs/cgroup` for each
    PE service.

  - The support script now collects the output of `systemctl status` for each
    PE service.

  - The support script now supports RedHat 8 operating systems.

### Bug Fixes

  - The support script no longer creates a login shell when switching users
    to run diagnostics.


## [Release 2.10.0] - 2019-01-15

### Summary

Feature and bugfix release of the Support Script.

### Features

  - The support script now gathers node activity counts from the
    PE Orchestration Service.

  - Windows event logs are now captured in CSV format and include
    timestamps.

### Bug Fixes

  - The output archives of the PowerShell and Ruby support scripts
    now match the naming convention used by the Bash script.


## [Release 2.9.1] - 2018-10-24

### Summary

Bugfix release of the Support Script.

### Bug Fixes

  - The `--log-age` flag is used to limit the age of metrics data collected
    from /opt/puppetlabs. This further reduces the size of support script
    archives by only collecting 14 days worth of metrics data.

  - The support script now consults `/etc/os-release` when detecting the
    which OS it is executing on before falling back to the `lsb_release`
    command or platform specific release files.

  - LDAP connection settings are collected directly from console-services.
    Previously, these were selected from the database, which required the
    script to be run on the DB node of a split install.

  - The script no longer calls `puppet master --configprint` to discover
    modulepath as this command has been removed from PE 2019.0.


## [Release 2.9.0] - 2018-09-25

### Summary

Feature and bugfix release of the Support Script.

### Features

  - Postgres HA replication status is collected from `pg_replication_slots`
    and `pg_stat_replication`.

  - Configuration and log files for the PE 2019.0 `pe-bolt-server` service
    are collected from `/etc/puppetlabs/bolt-server/conf.d` and from
    `/var/log/puppetlabs/bolt-server`.

  - On PE 2019.0 and newer, the `puppetserver ca` command is used instead
    of `puppet cert` to list issued certificates.

  - Windows agents now receive a `puppet-enterprise-support.ps1` PowerShell
    script that performs basic checks of agent services and gathers logs.

  - New Ruby version of the support script that can be selected by passing
    the `--v3` flag to `puppet enterprise support`. This feature is under
    development and the regular script should be preferred for production use.

### Bug Fixes

  - Remove the check for `puppet module changes` on modules bundled with PE.
    This check has terrible performace, is duplicated by the validation output
    from the pe-modules package, and fails on PE 2019.0 as the required
    checksums.json files are no longer present.

  - Test suite updates for PE 2019.0 and Puppet 6.


## [Release 2.8.0] - 2018-07-17

### Summary

Feature and bugfix release of the Support Script bundled with PE 2017.3.9
and PE 2018.1.3. See release notes for version 1.9.1 for a list of bugfixes.

### Features

  - Added support for Ubuntu 18.04.

  - Support script now collects the  output of `puppet infrastructure status`
    if the client-tools package is installed and configured. This is useful
    for checking the state of a HA deployment along with other infrastructure
    nodes hosting PE services.


## [Release 2.7.0] - 2018-04-13

### Summary


spec/spec_helper.rb
require 'logger'
require 'stringio'
require 'tmpdir'
require 'fileutils'


RSpec.configure do |c|
  c.mock_with :rspec
end

RSpec.shared_context 'stub script settings' do
  # Including this context creates a new settings instance before each test
  # instead of re-using a shared singleton instance. A logger that stores
  # data in a StringIO instance is also configured. The `noop` option is
  # also set to true in the default configuration so that external commands
  # triggered by unit tests are not executed.
  let(:script_settings) { PuppetX::Puppetlabs::SupportScript::Settings.new }
  let(:script_log) { StringIO.new }
  let(:script_logger) do
    logger = Logger.new(script_log)
    logger.level = Logger::DEBUG
    logger
  end
  let(:temp_drop_dir) { Dir.mktmpdir }


  before(:each) do
    allow(PuppetX::Puppetlabs::SupportScript::Settings).to receive(:instance).and_return(script_settings)
    script_settings.log.add_logger(script_logger)
    script_settings.configure(noop: true)
    script_settings.state[:drop_directory] = temp_drop_dir
  end

  after(:each) do
    script_logger.close
    FileUtils.remove_entry_secure(temp_drop_dir, force: true)
  end
end


require 'facter'
require 'puppet_x/puppetlabs/support_script/v3/puppet-enterprise-support.rb'


ext/run_acceptance_tests.sh
#!/bin/bash

PE_TEST_SERIES="2018.1"
TEST_MATRIX=()
PRESERVE_BEHAVIOR="never"

print_usage() {
  cat <<-EOF
USAGE: ./run_acceptance_tests.sh [-h] [-p preserve] [-r pe_series] [-t hostspec]

A small shell script that spawns a beaker test for each PE
infrastructure platform, in parallel, then launches a
webserver to display the test results.

  -h  Print help.
  -p  Change whether or not hosts are detroyed after tests run.
      Default value is "never". Other values are:
        always onpass onfail
  -r  Release series of PE to test against as an X.Y string.
      Also accepts the string "master" to test against the
      very latest build.
      Default value is: ${PE_TEST_SERIES}
  -t  A beaker-hostgenerator string to run tests against.
      May be passed multiple times.

For example, to run tests against the lastest "master" build of PE, with
monolithic installs of CentOS 7 and Ubuntu 18.04, and to preserve hosts
if tests fail:

./run_acceptance_tests.sh \\
  -p onfail \\
  -t centos7-64mdca -t ubuntu1804-mdca \\
  -r master

EOF
}

while getopts hp:r:t: flag; do
  case "${flag}" in
    h)
      print_usage
      exit 0
      ;;
    p)
      PRESERVE_BEHAVIOR="${OPTARG?}"
      ;;
    r)
      PE_TEST_SERIES="${OPTARG?}"
      ;;
    t)
      TEST_MATRIX+=("${OPTARG?}")
      ;;
    ?)
      print_usage
      exit 1
      ;;
  esac
done

if [[ "${#TEST_MATRIX[@]}" -eq 0 ]]; then
  # Default for the "master" branch, or releases since PE 2019.2.
  TEST_MATRIX=('centos7-64mdca'
               'centos8-64mdca'
               'sles12-64mdca'
               'ubuntu1604-64mdca'
               'ubuntu1804-64mdca'
               'centos7-64amdc-64compile_master.af-64agent%2Cpe_postgres.')

  # Special case older release pipelines.
  case "${PE_TEST_SERIES}" in
  '2018.1')
    TEST_MATRIX=('centos6-64mdca'
                 'centos7-64mdca'
                 'sles12-64mdca'
                 'ubuntu1604-64mdca'
                 'ubuntu1804-64mdca'
                 'centos7-64am-64ad-64ac-64compile_master.af')
    ;;
  '2016.4')
    TEST_MATRIX=('centos6-64mdca'
                 'centos7-64mdca')
    ;;
  esac
fi

build_url="https://artifactory.delivery.puppetlabs.net/artifactory/generic_enterprise__local/${PE_TEST_SERIES}/ci-ready/LATEST"
printf 'Reading latest good build from: %s\n' "${build_url}"

LATEST_GOOD_BUILD=$(curl -Ss --max-time 5 "${build_url}")

if [[ $? -ne 0 ]]; then
  printf '\nCould not download build information from:\n\t%s\nAre you connected to the internal Puppet network?\n' \
         "${build_url}" >&2
  exit 1
fi

printf "Testing build: %s\n" "${LATEST_GOOD_BUILD?}"

export BEAKER_PE_DIR="https://artifactory.delivery.puppetlabs.net/artifactory/generic_enterprise__local/${PE_TEST_SERIES}/ci-ready/"
export BEAKER_PE_VER="${LATEST_GOOD_BUILD}"

execute_beaker() {
  # Changing preserve-hosts from "never" to "onfail" will leave VMs behind for debugging.
  bundle exec beaker \
    --preserve-hosts "${PRESERVE_BEHAVIOR}" \
    --config "$1" \
    --debug \
    --keyfile ~/.ssh/id_rsa-acceptance \
    --pre-suite tests/beaker/pre-suite \
    --tests tests/beaker/tests | \
  grep 'PE-' | while read line; do
    printf '%s: %s\n' "${1}" "${line}"
  done
}

pids=()
for config in "${TEST_MATRIX[@]}"; do
  printf 'Spawning test for: %s\n' "${config}"
  execute_beaker "${config}" &
  pids+=("$!")
done

for pid in "${pids[@]}";do
  wait "${pid}"
done

pushd junit &> /dev/null || exit 1
bundle exec ruby -run -e httpd -- --bind-address=127.0.0.1 --port=8000


tests/beaker/tests/params.rb
test_name 'Params' do
  step 'PE-20886 - - Test directory parameter' do
    hosts.each do |host|
      result = on(host, puppet('enterprise support --dir /tmp', 'ENV' => { 'BEAKER_TESTING' => '1' }))
      directory = result.stdout.match(%r{^Support data is located at (.*)\/puppet_enterprise_support}).captures.first
      assert_match(%r{\/tmp}, directory, 'Path should begin with /tmp as specified by --dir /tmp')
    end
  end

  step 'PE-19805 -- Test ticket parameter' do
    hosts.each do |host|
      result = on(host, puppet('enterprise support --ticket 12345', 'ENV' => { 'BEAKER_TESTING' => '1' }))
      ticket = result.stdout.match(%r{^Support data is located at \/var\/tmp\/puppet_enterprise_support_(.*)_}).captures.first
      assert_match(%r{12345}, ticket, 'Path should include 12345 as specified by --ticket 12345')
    end
  end

  step 'PE-2736 -- Test upload parameter' do
    hosts.each do |host|
      target = 'customer-support.puppetlabs.net'
      # customer-support.puppetlabs.net does not resolve internally.
      on(host, "echo '10.230.16.41 #{target}' >> /etc/hosts")
      result = on(host, puppet('enterprise support --ticket BEAKER_TESTING --v3 --only base-status --upload'))
      server = result.stdout.match(%r{File uploaded to: (.*)}).captures.first
      assert_equal(server, target, "File should be uploaded to #{target} as specified by --upload")
    end
  end
end


lib/puppet_x/puppetlabs/support_script/v1/puppet-enterprise-support.ps1
################################################################################
# Parameters
################################################################################

Param (
  [int]$logAge = 14
 )

$script_version = '2.13.2'

################################################################################
Write-Host
Write-Host ('Puppet Enterprise Windows Support Script v' + $script_version)
Write-Host
################################################################################

################################################################################
# PowerShell Version Check
################################################################################

if ($PSVersionTable.PSVersion.Major -lt 3) {
  Write-Host 'Error: This script requires PowerShell Version 3 or newer.'
  Write-Host
  exit 1
}

Write-Host 'Note: This script may take a few minutes to execute.'
Write-Host

################################################################################
# Functions
################################################################################

Function Test-CommandExists {
 Param ($command)
 $eap = $ErrorActionPreference
 $ErrorActionPreference = 'stop'
 $exists = $FALSE
 Try {
    if (Get-Command $command){
      $exists = $TRUE
    }
  } Catch {
    "$command does not exist"
  }
  Finally {
    $ErrorActionPreference = $eap
  }
  return $exists
}

################################################################################
# Support Script Variables
################################################################################

$run_date_time = (Get-Date -Format 'yyyyMMddHHmmss')
$hostname = $env:computername.ToLower()
$time_zone = [System.TimeZone]::CurrentTimeZone
$eventlog_date = (Get-Date).AddDays(-$logAge)

if (Test-CommandExists ('puppet.bat')) {
  $puppet_conf               = [string](puppet.bat config print config)
  $puppet_server             = [string](puppet.bat config print server)
  $puppet_logdir             = [string](puppet.bat config print logdir)
  $puppet_statedir           = [string](puppet.bat config print statedir)
} else {
  $puppet_conf     = ''
  $puppet_server   = ''
  $puppet_logdir   = [string](Get-Location)
  $puppet_statedir = ''
}

if (Test-CommandExists ('facter.bat')) {
  $puppet_pxp_logdir         = [string](facter.bat -p common_appdata) + "\PuppetLabs\pxp-agent\var\log"
  $puppet_mcollective_logdir = [string](facter.bat -p common_appdata) + "\PuppetLabs\mcollective\var\log"
} else {
  $puppet_pxp_logdir         = ''
  $puppet_mcollective_logdir = ''
}

$output_directory = $puppet_logdir    + '/puppet_enterprise_support_' + $hostname + '_' + $run_date_time
$output_file      = $output_directory + '/support_script.log'
$output_archive   = $puppet_logdir    + '/puppet_enterprise_support_' + $hostname + '_' + $run_date_time + '.zip'

################################################################################
# PowerShell Variables
################################################################################

$global:progressPreference = 'SilentlyContinue'

################################################################################
# Output Validation
################################################################################

$(New-Item -Path $output_directory -ItemType directory) | Out-Null

@{version = $script_version; timestamp = $run_date_time; osfamily = 'windows'} | ConvertTo-Json | Out-File -FilePath ($output_directory + '/metadata.json')

if (! (Test-Path $output_directory)) {
  Write-Host 'Error: could not create output directory:'
  Write-Host $output_directory
  Exit
}

# For more information about redirection, see:
# https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_redirection
# Using *>> requires PowerShell 3.0 just to compile, but captures streams 3-6 (Warning, Verbose, Debug, Information).

'Puppet Enterprise Windows Support Script' > $output_file

if (! (Test-Path $output_file)) {
  Write-Host 'Error: could not write to output file:'
  Write-Host $output_file
  Exit
}

################################################################################
Write-Host 'Collecting Puppet Diagnostic Information ...'
################################################################################

'Hostname'              | Out-File -Append -FilePath $output_file
(hostname)              | Out-File -Append -FilePath $output_file

'Date, Time, Time Zone' | Out-File -Append -FilePath $output_file
$run_date_time          | Out-File -Append -FilePath $output_file
$time_zone.StandardName | Out-File -Append -FilePath $output_file

'Environment Path'      | Out-File -Append -FilePath $output_file
$env:path               | Out-File -Append -FilePath $output_file

'Get-Command puppet'    | Out-File -Append -FilePath $output_file
Get-Command puppet      | Out-File -Append -FilePath $output_file

'Get-Command facter'    | Out-File -Append -FilePath $output_file
Get-Command facter      | Out-File -Append -FilePath $output_file

'Get-Command ruby'      | Out-File -Append -FilePath $output_file
Get-Command ruby        | Out-File -Append -FilePath $output_file

'puppet.bat --version'  | Out-File -Append -FilePath $output_file
(puppet.bat --version)  | Out-File -Append -FilePath $output_file

'facter.bat --version'  | Out-File -Append -FilePath $output_file
(facter.bat --version)  | Out-File -Append -FilePath $output_file

'ruby --version'        | Out-File -Append -FilePath $output_file
(ruby --version)        | Out-File -Append -FilePath $output_file

################################################################################
Write-Host 'Running Facter in Debug Mode ...'
################################################################################

facter.bat --debug --trace 2>&1 | Out-File -Append -FilePath ($output_directory + '/facter.log')

################################################################################
Write-Host 'Testing Puppet Server Connectivity ...'
################################################################################

'Test Port 8140 to Puppet Server ' + $puppet_server  | Out-File -Append -FilePath $output_file
Test-NetConnection $puppet_server -Port 8140         | Out-File -Append -FilePath $output_file

'Test Port 8142 to Puppet Server ' + $puppet_server  | Out-File -Append -FilePath $output_file
Test-NetConnection $puppet_server -Port 8142         | Out-File -Append -FilePath $output_file

'Test Port 61613 to Puppet Server ' + $puppet_server | Out-File -Append -FilePath $output_file
Test-NetConnection $puppet_server -Port 61613        | Out-File -Append -FilePath $output_file

################################################################################
Write-Host 'Querying Puppet Agent Services ...'
################################################################################

'Puppet Agent Services Query: puppet pxp-agent mcollective' | Out-File -Append -FilePath $output_file
Get-WmiObject -Query "SELECT * FROM win32_service where (name = 'puppet' or name = 'pxp-agent' or name = 'mcollective')" | Format-List -Property Name, Pathname, ProcessId, Startmode, State, Status, Startname | Out-File -Append -FilePath $output_file

################################################################################
Write-Host 'Exporting Puppet Agent Services Event Logs ...'
################################################################################

# NOTE: This can be replaced with Get-WinEvent once Server 2008
#       is no longer supported. This will improve efficieny and
#       allow access to more event fields, such as process ID.
Get-Eventlog -Source puppet -LogName Application -After $eventlog_date 2> ($output_directory + '/eventlog_application_puppet_errors.txt')    | `
  Select-Object -Property @{label='TimeGenerated';expression={$_.TimeGenerated.ToString("o")}},Index,EntryType,Message| `
  Export-Csv -NoTypeInformation -Path ($output_directory + '/eventlog_application_puppet.csv')
Get-Eventlog -Source pxp-agent -LogName Application -After $eventlog_date 2> ($output_directory + '/eventlog_application_pxp-agent_errors.txt') | `
  Select-Object -Property @{label='TimeGenerated';expression={$_.TimeGenerated.ToString("o")}},Index,EntryType,Message| `
  Export-Csv -NoTypeInformation -Path ($output_directory + '/eventlog_application_pxp-agent.csv')

################################################################################
Write-Host 'Copying Puppet Agent Services Configuration Files and State Directory ...'
################################################################################

if (! (Test-Path $puppet_conf)) {
  'Error: puppet config file not found' | Out-File -Append -FilePath $output_file
} else {
  Copy-Item $puppet_conf -Destination $output_directory
}

if (! (Test-Path $puppet_statedir)) {
  'Error: puppet state directory not found' | Out-File -Append -FilePath $output_file
} else {
  Copy-Item $puppet_statedir -Recurse -Destination $output_directory
}

if (! (Test-Path $puppet_pxp_logdir)) {
  'Error: puppet pxp-agent log directory not found' | Out-File -Append -FilePath $output_file
} else {
  Copy-Item $puppet_pxp_logdir -Recurse -Destination ($output_directory + "\pxp-agent")
}

if (! (Test-Path $puppet_mcollective_logdir)) {
  'Error: puppet mcollective directory not found' | Out-File -Append -FilePath $output_file
} else {
  Copy-Item $puppet_mcollective_logdir -Recurse -Destination ($output_directory + "\mcollective")
}

################################################################################
Write-Host 'Compressing Support Script Data ...'
################################################################################

Add-Type -Assembly 'System.IO.Compression.FileSystem' ;
[System.IO.Compression.ZipFile]::CreateFromDirectory($output_directory, $output_archive);
Remove-Item $output_directory -Recurse

################################################################################
Write-Host
Write-Host 'Done.'
Write-Host
Write-Host 'Puppet Enterprise Windows Support Script output is located in:'
Write-Host
Write-Host $output_archive
Write-Host
Write-Host 'Please submit it to Puppet Enterprise Support.'
Write-Host
################################################################################


lib/puppet/face/enterprise/support.rb
require 'puppet/indirector/face'
require 'puppet/feature/base'

require 'open3'
require 'timeout'

# Primum non nocere

Puppet::Face.define(:enterprise, '1.0.0') do
  action :support do
    summary 'Collects Puppet Enterprise Support Diagnostics'

    # See also: lib/puppet_x/puppetlabs/support_script/v3/puppet-enterprise-support.rb

    default_dir     = File.directory?('/var/tmp') ? '/var/tmp' : '/tmp'
    default_log_age = 14

    supported_platforms = %w[redhat debian suse]

    option '--classifier' do
      summary 'Include Classifier data'
    end

    option '--dir DIRECTORY' do
      summary "Output directory. Defaults to: #{default_dir}"
      default_to { default_dir }
    end

    option '--encrypt' do
      summary 'Encrypt output using GPG'
    end

    option '--filesync' do
      summary 'Include FileSync data. Requires the --v3 parameter'
    end

    option '--log-age DAYS' do
      summary "Log age (in days) to collect. Defaults to: #{default_log_age}"
      default_to { default_log_age }
    end

    option '--ticket NUMBER' do
      summary 'Support ticket number'
    end

    option '--upload' do
      summary 'Upload to Puppet Support via SFTP. Requires the --ticket and --v3 parameters'
    end

    option '--upload-disable-host-key-check' do
     summary 'Disable SFTP Host Key Check. Requires the --upload parameter'
    end

    option '--upload-key FILE' do
      summary 'Key for SFTP. Requires the --upload parameter'
    end

    option '--upload-user USER' do
      summary 'User for SFTP. Requires the --upload parameter'
    end

    option '--v3' do
      summary 'Use Version 3.0 of this command (experimental)'
    end

    option '--enable LIST' do
      summary 'Comma-delimited list of scopes or checks to enable. Requires the --v3 parameter.'

      before_action do |_, _, options|
        if options.key?(:enable)
          # Copied from Ruby's OptionParser handler for Array values
          options[:enable] = options[:enable].split(',').collect {|ss| ss unless ss.empty?}
        end
      end
    end

    option '--disable LIST' do
      summary 'Comma-delimited list of scopes or checks to disable. Requires the --v3 parameter.'

      before_action do |_, arg, options|
        if options.key?(:disable)
          options[:disable] = options[:disable].split(',').collect {|ss| ss unless ss.empty?}
        end
      end
    end

    option '--only LIST' do
      summary 'Comma-delimited list of of scopes or checks to run, disabling all others. Requires the --v3 parameter.'

      before_action do |_, arg, options|
        if options.key?(:only)
          options[:only] = options[:only].split(',').collect {|ss| ss unless ss.empty?}
        end
      end
    end

    option '--list' do
      summary 'List available scopes and checks that can be passed to --enable, --disable, or --only. Requires the --v3 parameter.'
    end

    when_invoked do |options|
      support_module = File.expand_path(File.join(File.dirname(__FILE__), '../../../..'))

      if Puppet.features.microsoft_windows?
        support_script = File.join(support_module, 'lib/puppet_x/puppetlabs/support_script/v1/puppet-enterprise-support.ps1')
        begin
          Open3.popen2e('powershell.exe', '-File', support_script) do |_i, oe, t|
            begin
              Timeout.timeout(120) do
                puts oe.readline until oe.eof?
              end
            rescue Timeout::Error
              Process.kill('KILL', t.pid)
              Puppet.err('The powershell command timed out.')
              exit 1
            end
          end
        rescue => e
          Puppet.err("The powershell command returned: #{e.message}")
          exit 1
        end
        exit 0
      end

      os_family = `/opt/puppetlabs/puppet/bin/facter os.family`.chomp.strip
      unless supported_platforms.include? os_family.downcase
        Puppet.err("This command is not implemented for #{os_family} platforms at this time.")
        exit 1
      end

      support_script_parameters = []

      if options[:classifier]
        support_script_parameters.push('-c')
      end

      unless options[:dir] == ''
        support_script_parameters.push("-d#{options[:dir]}")
      end

      if options[:encrypt]
        support_script_parameters.push('-e')
      end

      if options[:filesync] && options[:v3] != true
        Puppet.err('The filesync parameter requires the --v3 parameter.')
        exit 1
      end

      if options[:log_age].to_s =~ %r{^\d+|all$}
        support_script_parameters.push("-l#{options[:log_age]}")
      else
        Puppet.err("The log-age parameter must be a number, or the string 'all'. Got: #{options[:log_age]}")
        exit 1
      end

      if options.key?(:ticket)
        if options[:ticket] =~ %r{^[\d\w\-]+$}
          support_script_parameters.push("-t#{options[:ticket]}")
        else
          Puppet.err "The ticket parameter may contain only numbers, letters, and dashes. Got: #{options[:ticket]}"
          exit 1
        end
      end

      if options[:upload] && (options[:v3] != true)
        Puppet.err('The upload parameter requires the --v3 parameter.')
        exit 1
      end

      [:enable, :disable, :only, :list].each do |opt|
        if options.key?(opt) && (options[:v3] != true)
          Puppet.err('The --%{opt} parameter requires the --v3 parameter.' %
                     {opt: opt})
          exit 1
        end
      end

      if options[:v3]
        require 'puppet_x/puppetlabs/support_script/v3/puppet-enterprise-support'
        PuppetX::Puppetlabs::SupportScript::Settings.instance.configure(**options)
        PuppetX::Puppetlabs::SupportScript::Settings.instance.log.add_logger(PuppetX::Puppetlabs::SupportScript::LogManager.console_logger)
        support = PuppetX::Puppetlabs::SupportScript::Runner.new
        support.add_child(PuppetX::Puppetlabs::SupportScript::Scope::Base, name: '')
        support.run
        return
      else
        support_script = File.join(support_module, 'lib/puppet_x/puppetlabs/support_script/v1/puppet-enterprise-support.sh')
        Kernel.exec('/bin/bash', support_script, *support_script_parameters)
      end
    end
  end
end
